### DISCUSSION TITLE
RSC and CDN cache problems makes the next.js useless for the highload projects

### DISCUSSION THREAD

**Discussion by maslowivan** [vercel/next.js#59167] [App Router]
Everytime user moves to the dynamic page and see the links, RSC makes different /xxx?_rsc=yyy requests with different _rsc hash. E.g. you have the e-commerce shop with the categories and the products.

Those _rsc requests have absolutely no cache, and CDN ignores them and pass them directly to the server, even we have the full page cached on the CDN for the specific product and the category.

_And we even have to have the custom server, because we can't pass the headers with the cache-tag for the products to the CDN.
We can't have cache-tag working like this /myproduct, because we have a lot of countries and the normally the cache-tag should be like product-123456 (by id) so we can drop it by the product id but that is another problem._

(Next.js also slow on the cloudflare pages when we test, so aws look the only option at the moment.)

### Problem example

from start page (see 71j4w)

**Next-Router-State-Tree** header: `["",{"children":[["lang","se","d"],{"children":["(defaultLayout)",{"children":["__PAGE__",{}]}]},null,null,true]}]`

<img width="1316" alt="image" src="https://github.com/vercel/next.js/assets/1205208/50dffc37-e72b-4ac9-8e5c-f17665f4d69c">

from another page (now is 4barx)

**Next-Router-State-Tree** header: `["",{"children":[["lang","se","d"],{"children":["(defaultLayout)",{"children":[["slug","dope","d"],{"children":["__PAGE__?{\\"gender\\":\\"Man\\"}",{}]}]}]},null,null,true]}]`

<img width="1227" alt="image" src="https://github.com/vercel/next.js/assets/1205208/b9a80f6c-5cb1-4166-9ea8-d43bdf254d10">

### Why do we need that?

Why can't we have the entire page passed by the `?_rsc=1` request with some configuration in the next.config file, to opt out the RSC cache system based on the router state but make it always return all the components?
It must always be ?_rsc=1 with some configuration - so doesn't matter from what page I did open it and it would make possible to **CACHE it on the CDN level**, because the APP level is slow.

CDN level is close to the user and it makes the dynamic pages won't trigger the database requests again, also it makes possible to have server only for the static CDN generation.

What is actually difference for me between the 71j4w and 4barx? Why do I need to make the new request to the server?

To handle million requests with the 100k+ products we should really have a lot of servers count with the very expensive configuration. Vercel is much slower here and more expensive than aws or even aws wrappers like flightcontrol - so we have to think here about the servers.

I have not so many solutions right now to make it works

Faster solutions:
1. keep all as is - let the server handle these requests, quickest solution
2. run cache for every _rsc request and cache everything - check how many cache entries do we produce and does it hurt the server or not
3. cache rsc partically somehow

Crazy and complex solutions:
1. make CDN worker run this logic (requests with ?_rsc) and make it’s own inner cache for the rsc, working with the header Next-Router-State-Tree
2. hack the Link component or the fetch server response file https://github.com/vercel/next.js/blob/87e0b4495a2710a7851a14430738519f27b2ebe4/packages/next/src/client/components/router-reducer/fetch-server-response.ts#L102

Maybe somebody knows something...

**Comment by josetr** [comment]
Can someone explain to me why these RSC requests are used even when every single layout / page is a Client Component?

layout.tsx : 
```ts
"use client"

export default function Layout({children}) => children
```

page1/page.tsx : 
```ts
"use client"
import { Link } from "next/link"

export default function Page() => <Link href="/page2">page 2</Link>
```

page2/page.tsx : 
```ts
"use client"
import { Link } from "next/link"

export default function Page() => <Link href="/page1">page 1</Link>
```

**Reply by ImanAkrami** [reply]
Thanks for asking this. I am baffled by this too.

**Reply by amanmamgain9** [reply]
My understanding is even if you use "use client" the shell will be server rendered...then for the new page you will have the js bundle be part of that shell then that will load and finally your page will render

**Comment by vitalyiegorov** [comment]
@studentIvan can you please describe how do you use cdn infront of the nextjs server?

**Reply by ats1999** [reply]
Which cdn you want to use?

**Reply by maslowivan** [reply]
@vitalyiegorov we have our own infrastructure on Amazon AWS + cloudflare CDN infront

**Comment by josetr** [comment]
Also, regarding Cloudflare:

`@cloudflare/next-on-pages` will generate a [_routes.json](https://developers.cloudflare.com/pages/platform/functions/routing/#create-a-_routesjson-file) file which is not very optimal 

`{"version":1,"description":"Built with @cloudflare/next-on-pages@1.8.2.","include":["/*"],"exclude":["/_next/static/*"]}`

This means a worker will run for all paths except `/next_static/*`, so even if you have existing `.rsc` files, they will still cost you one serverless function execution.

But in my previous example, even when all pages are Client Components, and those .rsc files actually exist, they are still ignored, and a `page1?rsc=` request is made, even tho a `page1.rsc` request would suffice.

Deploying my previous example to Vercel doesn't seems very optimal either, because at `vercel.com/org/project/deployment-id/functions` I see 3 functions for my actual pages, and 3 that handle those `RSC` requests.

If someone is on a tight budget, stick to the Pages router if you can.

**Reply by kumar0** [reply]
Any idea how do we implement in cloudfront.. took lot of effort to move to approuter.. once all done.. stuck with this page refresh issue..

**Comment by amannn** [comment]
Same problem, this is quite an issue for me for a high-traffic website.

E.g.:

```js
// /?_rsc=gepvv
/products/1 → / 

// /?_rsc=11gmh
/products/2 → / 
```

… although the RSC payload returned from the request is identical. This results in a huge number of different URLs that need to be saved in the CDN cache.

Imagine a large e-commerce app that shows related products on a product detail page now having to cache all permutations of links between products instead of one page per product. Not sure if I'm missing something, but this seems to be quite a big issue for high-traffic sites that use dynamic segments with a lot of possible values. Cache size is one aspect here but the amount of requests that need to be handled by the origin server is another one.

I'm probably missing a lot of details here, but I'm wondering if this could be solved at:

https://github.com/vercel/next.js/blob/87e0b4495a2710a7851a14430738519f27b2ebe4/packages/next/src/client/components/router-reducer/fetch-server-response.ts#L81-L86

… if the hash would only consider the currently active parent layout segments? 

E.g.:
```tsx
// ?_rsc=hash('products')
/products/1 → /

// … or a more complex example:

// ?_rsc=hash('en-us', 'products', 1, 'comments')
/en-us/products/1/comments/2 → /product/2
```

In these examples, the last part of the URL is not included in the RSC hash. Wouldn't that transport all relevant details, i.e. which layouts are already fetched, to efficiently fetch the minimum markup for the target page?

Depending on the structure of an app, it might even be more efficient to simply fetch the whole markup for any given target page instead of granular cache entries that consider the current page if you have many pages below dynamic segments (e.g. `/comments`, `/details`, …).

Note that this wasn't an issue with the Pages Router as every page downloaded a cacheable JSON object that didn't depend on the previous route.

FWIW, Remix doesn't seem to have this issue with their implementation of nested routes. They don't support RSC yet, but it seems like the problem is more related to nested routes than to RSC as far as I can tell.

@huozhi any chance you could comment on this as you've worked on https://github.com/vercel/next.js/pull/50970?

**Comment by ImanAkrami** [comment]
are there really no solution except caching the requests on CDN? this was the initial reason I've move FROM pages router, to not have those mandatory "getServerSideProps" where they are identical.

It seems very weird that as @josetr  mentioned, a fully "use client"ed app would still make server requests.

also, I believe these are related

#60412 
#60951

**Comment by dankain** [comment]
Hi I'm trying to work out how I make the RSC requests work with a CDN when self hosting Nextjs. Is there any more info on this subject. I'm working on an ecommerce site with 100,000 products. Those products could appear in numerous product listing pages (search result pages). Each of those PLP pages are given unique URLs for SEO purposes. Take for example the following URLs:

`/mens/`
`/mens/trainers`
`/mens/trainers/brand`
`/mens/trainers/brand?facet-price=%3A168`

They could all have the same products in. Due to the way the rsc hash is calculated that means I get a different `_rsc` params on each listing page, even though the contents of that response is exactly the same.

`product/299336/?_rsc=1vl30`
`product/299336/?_rsc=qe3go`
`product/299336/?_rsc=1vg99`
`product/299336/?_rsc=1stsw`

I have even gone to different areas of the site cart, checkout, order history, all with links back to the same product, they each produce different  `_rsc` params, but still the data returned is identical?

On a high throughput site I want to be able to cache identical data in the CDN close to the customer. At the moment that would be impossible as there would be to many variations on the rsc hash to make the caching effective.

For solutions I think I only have 2 options:

- **Cache all rsc requests in the CDN**  - this would end up caching loads of duplicate data and get cache misses when they should be hits
- **Pass all request through to the Nextjs server.** - With this solution I would worry the server would be overloaded at peak periods.

In both cases there would be an extra cost to the client

I'm trying to understand why the rsc has is different when it is always returning identical data? What is the purpose of this hash? As mentioned before could this just be set to `_rsc=1`? We would also have issues with the Vary header as it is currently returned like this:

`Vary:  RSC, Next-Router-State-Tree, Next-Router-Prefetch, Next-Url`

In this case the Next-Router-State-Tree and Next-Url will be set on where you are coming from and do not necessarily have an impact on the data need for the page we are going to. The Vary header will again have an impact on the CDN

**Comment by barroudjo** [comment]
This `_rsc` query parameter is also used in next static builds (`output = 'export'`), which makes no sense since the build is static and there is only one rsc (with a `.txt` file extension) file per url...
Really this code makes very little sense, and seems to create more problems than solutions.

**Reply by arfa123** [reply]
I am building full offline support web app and this `_rsc=` making problem for me to cache all routes. Can anyone have solution for this?

**Comment by eric-burel** [comment]
Hey folks, I feel like you might want to open an actual bug ticket if you think the current behavior is not reasonable, with a minimal reproduction of the caching issue you experience. It's hard to reason based only on the messages sent here even for experienced Next.js dev I think. This feels quite advanced and might require changes in the way Next.js works so I am not sure you'll get the answers you want in a discussion. Unless this has already be moved from a discussion ?

**Reply by dankain** [reply]
As suggested I have raised an issue https://github.com/vercel/next.js/issues/65335

**Comment by fuma-nama** [comment]
Was facing this problem on my projects.

RSC payload is used to load pages after initial load, so we don’t have to load the entire HTML document. 
However, the RSC request uses serverless functions (on Vercel) instead of CDN, this is happening even on static pages.

My understanding is that client-side navigation (e.g. Next.js Link) doesn’t know whether the page is rendered on-demand. To support dynamic rendering, they made it a serverless function to ensure users are fetching the latest info from server, which also impacts the static pages. Hence, the RSC payload is **not cached by browser**, and is **not served via CDN**.

This means if the user is not opening the static page directly (on initial load), opening a page is almost same as requesting a server rendered route. On Vercel, this caused a huge usage of serverless functions and data cache read (because it is cached by Vercel Data Cache), especially with prefetching enabled on the Next.js Link component.

I cannot think of an optimal solution, the only idea is to use static export so at least it won't cause a huge usage of serverless functions.

**Comment by fmnxl** [comment]
RSC payload not being cached by the CDN nor by Next's caching system means interception routes (and parallel routes) are broken after deployments unless the whole cache is cleared.

I implemented a modal with interception + parallel route, which breaks down into a full page load because the RSC payload hash has changed, and the old hash is no longer served by the new build.

To keep the modal working, I have been clearing the (CDN) cache after each deployment, which hurts load time.

**Comment by raine** [comment]
When using static export, what are the appropriate cache headers for these `.txt?rsc=1iwkq` requests?

**Comment by jaan1025** [comment]
Still seeking an explanation in 2025.

**Reply by amannn** [reply]
There was an update on this issue here: https://github.com/vercel/next.js/issues/65335#issuecomment-2583401824

**Comment by charlie0077** [comment]
Can next.js team just provide a way to disable this _rsc feature?
I built a SPA, and now when user click some button that triggers router.push(xxx), sometimes they will need to wait for 0.5s for server response, which is ridiculous. 

I think the default setting should work for SPA, things like this should be a opt-in.

**Reply by sudo-iudo** [reply]
Disable Next.js, enable Remix 👍

**Reply by VladBo** [reply]
Replacing `next/link` with `<a>` tag seems like the only way to avoid having `_rsc=` in the URL and use CDN caching instead.

**Reply by lucasbasquerotto** [reply]
I created a Feature Request for that https://github.com/vercel/next.js/discussions/83184, but I'm not sure if the NextJS team would be interested in that, as more server calls (from those using the Vercel platform) means more money for them. I would use some alternative framework instead of NextJS if there was a strong choice that offered seamless SSR in the first call (for SEO and OG metadata), and full client side navigation from that point on. 

I will probably take a look at Remix if NextJS takes too much time to offer a full SPA option after the 1st page load (I would expect such a feature to be day-0 for any React framework that does SSR), otherwise our users would probably have a slow experience, especially considering that our pages are dynamic (we use strict CSP for scripts for security), and the subsequent server RSC calls are entirely pointless after the 1st page load.

**Comment by VishalKaleria** [comment]
any update on this?

**Reply by b2m9** [reply]
@VishalKaleria There was an update in January that they're working on it: https://github.com/vercel/next.js/issues/65335#issuecomment-2583401824
Following this issue is probably a good idea.

**Reply by maslowivan** [reply]
I've solved it in our projects by myself, not sure how to make this solution global for anybody. You can check how it works here https://www.dopesnow.com

**Reply by madfcat** [reply]
@maslowivan so, what's your approach?

**Reply by maslowivan** [reply]
@madfcat will prepare some answer and let you know soon, thanks for pushing

**Comment by VishalKaleria** [comment]
How you did this?

**Reply by maslowivan** [reply]
I will prepare some good readable answer and post to this thread

**Reply by VishalKaleria** [reply]
Any updates???

**Reply by litianming1** [reply]
> I will prepare some good readable answer and post to this thread

Would you be able to share a repository or some references for this solution? I would be extremely grateful!

**Reply by tanwarvinay** [reply]
> I will prepare some good readable answer and post to this thread

Hey, looking forward to the solution you came up with.

I looked into https://www.dopesnow.com/ ,  noticed that the rsc calls are cached to cdn, most of rsc calls have =default, but some also have other hash like this one /orders?_rsc=createmen,

calls with hash as "default" were already cached, and had cf-cache-status as HIT, 

but request with "createmen" was not cached, only after 2 or 3 page reloads this request got cf-cache-status as HIT. Is there particular reason that it was not cached after the first request.

**Reply by maslowivan** [reply]
@tanwarvinay yep, different initial layout

**Reply by martenjurgens** [reply]
@maslowivan Have you posted the solution yet? We are having the same issue with our project and would like to solve it

**Comment by VishalKaleria** [comment]
Awesome, thank you so much!

**Reply by litianming1** [reply]
Hello, have you solved this problem in the project?
